### PR TITLE
Feat: Add support for custom query parameters @W-13974911@

### PR DESCRIPTION
**Pull request by joeluong-sfcc**
This PR adds support for custom query parameters as SCAPI now supports SCAPI hooks: https://developer.salesforce.com/docs/commerce/commerce-api/references/about-commerce-api/about.html#08302023

Query params that begin with `c_` are considered custom query parameters and are passed onto the underlying SCAPI call. Any other query params that aren't a part of the SCAPI endpoint spec will be filtered out and not passed.

This PR also adds node 18 & 20 to the CI

Testing this PR takes a bit of set up:

1. `git clone git@github.com:SalesforceCommerceCloud/commerce-sdk-isomorphic.git`
2. `git checkout ju/support-custom-params`
3. `yarn install`
4. modify `templates/operations.ts.hbs` to add a `console.log()` to check query params:
```javascript
// operations.ts.hbs
...
      console.log('URL', url.toString())
      const response = await fetch(url.toString(), requestOptions);
      if (rawResponse) {
        return response;
...
```
5. `yarn run renderTemplates && yarn run build:lib`
6. Create a new node project in a separate directory: 
```bash
mkdir testCustomParams
cd testCustomParams
npm init -y
touch index.js
```

7. modify `package.json` in `testCustomParams` directory to point `commerce-sdk-isomorphic` to local directory in `dependencies` section and add `"type": "module"`:

```json
  "type": "module",
  "dependencies": {
    "commerce-sdk-isomorphic": "<insert_path_to_local_commerce-sdk-isomorphic>"
  }
```

8. `npm install`
9. Add test code to `index.js` and replace client creds

```javascript
import pkg from 'commerce-sdk-isomorphic';
const { helpers, ShopperLogin, ShopperProducts } = pkg;

// demo client credentials, if you have access to your own please replace them below.
const CLIENT_ID = "<CLIENT_ID>";
const ORG_ID = "<ORG_ID>";
const SHORT_CODE = "<SHORT_CODE>";
const SITE_ID = "<SITE_ID>";

// must be registered in SLAS. On server, redirectURI is never called
const redirectURI = "http://localhost:3000/callback";

// client configuration parameters
const clientConfig = {
  parameters: {
    clientId: CLIENT_ID,
    organizationId: ORG_ID,
    shortCode: SHORT_CODE,
    siteId: SITE_ID,
  },
};

const slasClient = new ShopperLogin(clientConfig);

// GUEST LOGIN
const guestTokenResponse = await helpers
  .loginGuestUser(slasClient, { redirectURI })
  .then((guestTokenResponse) => {
    console.log("Guest Token Response: ", guestTokenResponse);
    return guestTokenResponse;
  })
  .catch((error) =>
    console.log("Error fetching token for guest login: ", error)
  );

const shopperProducts = new ShopperProducts({
    ...clientConfig,
    headers: {authorization: `Bearer ${guestTokenResponse.access_token}`}
});

const productsResult = await shopperProducts.getProducts({
    parameters: {
        ids: "25720044M,25686395M", 
        expand: ["promotions", "images", "prices"],
        c_customParamOne: '1',
        c_customParamTwo: 2,
        c_customParamThree: 3,
        invalidParam: 'hello',
    }
})

console.log("productsResult: ", productsResult)
```

10. `node index.js > out.txt`

11. Observe `out.txt` and that the custom query params are included in the fetch call and invalid query params are excluded from the call:
```
URL https://<short_code>.api.commercecloud.salesforce.com/product/shopper-products/v1/organizations/<org_id>/products?ids=25720044M%2C25686395M&expand=promotions%2Cimages%2Cprices&siteId=RefArch&c_customParamOne=1&c_customParamTwo=2&c_customParamThree=3
```

### CODE REVIEW

File: src/test/parameters.test.ts [hunk 1]
```diff
@@ -140,4 +140,40 @@ describe('Parameters', () => {
...
+        siteId: SITE_ID,
+        clientId: CLIENT_ID,
+        // expect `c_validCustomParam` but not `invalidParam`
+        c_validCustomParam: 'custom_param',
+      })
+      .reply(200, MOCK_RESPONSE);
+
+    const response = await customersClient.authorizeCustomer(options);
+
+    expect(response).toEqual(MOCK_RESPONSE);
```
Comment (johnboxall) [inline src/test/parameters.test.ts:177]: Nice test case!

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -9,7 +9,7 @@ jobs:
   linux-tests:
     strategy:
       matrix:
-        node: [10, 12, 14, 16]
+        node: [10, 12, 14, 16, 18]
```
Comment (johnboxall) [inline .github/workflows/test.yml:12]: Given that we've tentatively merged Node 20 support into PWA Kit @ 3.4, would you consider adding Node 20 here?

File: templates/operations.ts.hbs
Comment (johnboxall) [inline templates/operations.ts.hbs:1]: I can only guess at what we're doing here, but looks right 😅

**Review by johnboxall — APPROVED** [review]
Thanks @joeluong-sfcc – I ended up coming the behaviour by running HTTP requests through a MITM proxy:

```js
// clear ; NODE_EXTRA_CA_CERTS=~/.mitmproxy/mitmproxy-ca-cert.cer node index.js
import csi from "commerce-sdk-isomorphic";
const { helpers, ShopperLogin, ShopperProducts, ShopperBaskets } = csi;
import hpa from "https-proxy-agent";
const { HttpsProxyAgent } = hpa;

const redirectURI = "http://localhost:3000/callback";

const clientConfig = {
  parameters: {
    clientId: "",
    organizationId: "",
    shortCode: "",
    siteId: "",
  },
  fetchOptions: {
    agent: new HttpsProxyAgent("http://localhost:3001"),
  },
};

const guestTokenResponse = await helpers.loginGuestUser(
  new ShopperLogin(clientConfig),
  {
    redirectURI,
  }
);

const productsResult = await new ShopperProducts({
  ...clientConfig,
  headers: { authorization: `Bearer ${guestTokenResponse.access_token}` },
}).getProducts({
  parameters: {
    ids: "25720044M",
    c_custom: "custom",
    invalid: "123",
  },
});

console.log("productsResult: ", productsResult);
```

I can see that the outgoing requests includes the expected `c_custom` param. Yah!

File: .github/workflows/test.yml (hunk 2 above)
Comment (adamraya) [reply .github/workflows/test.yml:12]: Do we want to also include that info in the README? https://github.com/SalesforceCommerceCloud/commerce-sdk-isomorphic/blob/580fc611b22f96cb68830ec3e074f600e89e0f73/README.md#L11

**Review by adamraya — APPROVED** [review]
I followed the test steps and confirmed that the custom query parameters beginning with the prefix "c_" are included in the request, while the remaining query parameters are ignored.

But, I did notice an issue related to the addition of the new Node version https://github.com/SalesforceCommerceCloud/commerce-sdk-isomorphic/pull/139/files#r1451043899

File: templates/operations.ts.hbs
Comment (joeluong-sfcc) [reply templates/operations.ts.hbs:1]: If you run `yarn run renderTemplates` and then check the output directory (`src/lib/`), you can see the rendered typescript, probably easier to look at then the handlebar template

File: .github/workflows/test.yml (hunk 2 above)
Comment (joeluong-sfcc) [reply .github/workflows/test.yml]: I've added node 20 to the CI and updated the `README` to include Node `^18.x`.

@adamraya As for the `yarn install` error, this doesn't appear on my machine and the CI seems to be able to run `yarn install` successfully for both node 18 and 20: https://github.com/SalesforceCommerceCloud/commerce-sdk-isomorphic/actions/runs/7546319904/job/20543824549

What version of `yarn` do you have installed on your machine? Have you tried deleting `node_modules` and `yarn.lock` and then running `yarn install`?

**Review by adamraya — APPROVED** [review]

File: README.md [hunk 3]
```diff
@@ -8,7 +8,7 @@ The Salesforce Commerce SDK (Isomorphic) allows easy interaction with the B2C Co
 
 ### Requirements
 
-- Node `^12.x`, `^14.x`, or `^16.x`
+- Node `^12.x`, `^14.x`, `^16.x`, or `^18.x`
```
Comment (adamraya) [inline README.md:11]: nit: 
```suggestion
- Node `^12.x`, `^14.x`, `^16.x`, `^18.x` or `^20.x`

```

**Review by adamraya — APPROVED** [review]

**Pull request merged**